File: src/co/nums/intellij/aem/htl/file/HtlFileType.java
package co.nums.intellij.aem.htl.file;

import co.nums.intellij.aem.htl.HtlLanguage;
import co.nums.intellij.aem.htl.highlighter.HtlTemplateHighlighter;
import co.nums.intellij.aem.htl.icons.HtlIcons;
import com.intellij.lang.Language;
import com.intellij.openapi.fileTypes.FileTypeEditorHighlighterProviders;
import com.intellij.openapi.fileTypes.LanguageFileType;
import com.intellij.openapi.fileTypes.TemplateLanguageFileType;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;

import javax.swing.Icon;

public class HtlFileType extends LanguageFileType implements TemplateLanguageFileType {

	public static final HtlFileType INSTANCE = new HtlFileType();

	@NonNls
	public static final String DEFAULT_EXTENSION = "htl";

	private HtlFileType() {
		this(HtlLanguage.INSTANCE);
	}

	private HtlFileType(Language language) {
		super(language);

		FileTypeEditorHighlighterProviders.INSTANCE.addExplicitExtension(
				this,
				(project, fileType, virtualFile, editorColorsScheme) ->
						new HtlTemplateHighlighter(project, virtualFile, editorColorsScheme)
		);
	}

	@NotNull
	@Override
	public String getName() {
		return "HTL";
	}

	@NotNull
	@Override
	public String getDescription() {
		return "HTML Template Language file";
	}

	@NotNull
	@Override
	public String getDefaultExtension() {
		return DEFAULT_EXTENSION;
	}

	@Override
	public Icon getIcon() {
		return HtlIcons.HTL_FILE;
	}

}

File: src/co/nums/intellij/aem/htl/file/HtlFileType.kt
package co.nums.intellij.aem.htl.file

import co.nums.intellij.aem.htl.HtlLanguage
import co.nums.intellij.aem.htl.highlighter.HtlTemplateHighlighter
import co.nums.intellij.aem.htl.icons.HtlIcons
import com.intellij.openapi.fileTypes.FileTypeEditorHighlighterProviders
import com.intellij.openapi.fileTypes.LanguageFileType
import com.intellij.openapi.fileTypes.TemplateLanguageFileType

object HtlFileType : LanguageFileType(HtlLanguage), TemplateLanguageFileType {

    init {
        FileTypeEditorHighlighterProviders.INSTANCE.addExplicitExtension(
                this
        ) { project, fileType, virtualFile, editorColorsScheme -> HtlTemplateHighlighter(project, virtualFile, editorColorsScheme) }
    }

    override fun getName() = "HTL"

    override fun getDescription() = "HTML Template Language file"

    override fun getDefaultExtension() = "htl" // what about html?

    override fun getIcon() = HtlIcons.HTL_FILE

}
